### PR TITLE
Support for merged depth-stencil states

### LINKED ISSUE #376: vkCmdPipelineBarrier(): Image barrier 0x0x560107cfb730 references image 0x65 of formatVK_FORMAT_D24_UNORM_S8_UINT that must have the depth and stencil aspects set, but its aspectMask is 0x4.
I don't know where exactly this comes from but it's only since v 0.4 :)

VALIDATION [VUID-VkImageMemoryBarrier-image-01207 (0)] : vkCmdPipelineBarrier(): Image barrier 0x0x560107cfb730 references image 0x65 of formatVK_FORMAT_D24_UNORM_S8_UINT that must have the depth and stencil aspects set, but its aspectMask is 0x4. The Vulkan spec states: If image has adepth/stencil format with both depth and stencil components, then the aspectMask member of subresourceRangemust include both VK_IMAGE_ASPECT_DEPTH_BIT and VK_IMAGE_ASPECT_STENCIL_BIT (https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-VkImageMemoryBarrier-image-01207)
object info: (type: IMAGE, hndl: 101)

Best,
Yatekii

### PR DESCRIPTION
**Pull request by kvark**
Fixes #376 
@Yatekii would you be able to test it and confirm the fix?

### CODE REVIEW

**Comment by kvark** [comment]
@grovesNL this is largely rewritten/extended and should be ready now. Please have a look when you get a chance.

File: wgpu-core/src/track/texture.rs [hunk 1]
```diff
@@ -118,14 +117,12 @@ impl ResourceState for TextureState {
...
-                (hal::format::Aspects::DEPTH, &mut mip.depth),
-                (hal::format::Aspects::STENCIL, &mut mip.stencil),
-            ] {
-                if !selector.aspects.contains(aspect) {
+            for &mut (mip_aspects, ref mut plane_states) in mip {
+                let aspects = selector.aspects & mip_aspects;
+                if aspects.is_empty() {
                     continue;
                 }
+                debug_assert_eq!(aspects, mip_aspects);
```
Comment (grovesNL) [inline wgpu-core/src/track/texture.rs]: Why do we always expect this to be the case? Or is it just unhandled for now?

File: wgpu-core/src/track/texture.rs [hunk 2]
```diff
@@ -163,23 +160,8 @@ impl ResourceState for TextureState {
...
-                    &mip_other.depth,
-                ),
-                (
-                    hal::format::Aspects::STENCIL,
-                    &mut mip_self.stencil,
-                    &mip_other.stencil,
-                ),
-            ] {
+            for (&mut (aspects, ref mut planes_self), &(aspects_other, ref planes_other)) in mip_self.iter_mut().zip(mip_other) {
+                debug_assert_eq!(aspects, aspects_other);
```
Comment (grovesNL) [inline wgpu-core/src/track/texture.rs]: Same question here – why is this always true?

File: wgpu-core/src/track/texture.rs (hunk 1 above)
Comment (kvark) [reply wgpu-core/src/track/texture.rs]: We do expect this: the unit of tracking needs to be always smaller (or equal) than the requested set of aspects:
  - if either is `color`, there is no other way for the other to also be `color` (there aren't formats with both color and non-color aspectrs)
  - if the unit is a merged depth-stencil, we require the calling side (tracking code) to always request them both (see the change to `texture_create_view`)
  - if the unit is separate depth-stencil, then each is always going to be a subset of the requested set of aspects

Comment (grovesNL) [reply wgpu-core/src/track/texture.rs]: Alright this makes sense, thanks!

**Comment by kvark** [comment]
Bors r=grovesNL

**Pull request merged**